### ISSUE TITLE
Implement Beyond the Rainbow (BTR) Algorithm

### ISSUE THREAD

**Issue by jbuerman** [Stable-Baselines-Team/stable-baselines3-contrib#314]
I would like to contribute Beyond the Rainbow (BTR) to Stable-Baselines3 which improves over Rainbow Deep Q-Network (DQN) with six improvements from across the RL literature and is designed with computational efficiency in mind to train on high-end desktop PCs. 

Paper: https://arxiv.org/abs/2411.03820 

Code: https://github.com/VIPTankz/BTR 

**Background** 

Beyond the Rainbow (BTR) is an image-based RL algorithm with a discrete action space that improves over Rainbow DQN by adding 6 further improvements, namely, Impala (Scale=2), Adaptive Maxpooling (6x6), Spectral Normalization, Implicit Quantile Networks, Munchausen and Vectorized Environments. The algorithm has stated to gain traction (https://scholar.google.com/scholar?cites=3310089883274021659). 

BTR is competitive with recent algorithms like Dreamer-v3 (Hafner et al., 2023) or MEME (Kapturowski et al.,2023) considering its focus on training in more resource restricted environments like desktop PCs. The algorithm has been benchmarked by training on a high-end desktop PC, achieving a human-normalized interquartile mean (IQM) of 7.4 on Atari-60 within 12 hours. 

The implementation is based on PyToch. 

**Benefits** 

- Provide a state-of-the-art algorithm that provides the capability to train on high-end desktop which is of interest to smaller research labs and hobbyist who won’t have access to the hardware to train with more resource intensive algorithms.
- BTR can handle complex 3D games and has been used to train agents for Super Mario Galaxy, Mario Kart and Mortal Kombat (https://www.youtube.com/playlist?list=PL4geUsKi0NN-sjbuZP_fU28AmAPQunLoI) gaining interest from a community around building agents for games. 

**Practical Details** 

I will be working with the original author Tyler Clark to ensure that the SB3 implementation will achieve the performance stated in the paper.

**Comment by araffin** [comment]
Hello,
thanks for the proposal, but I would actually prefer to have rainbow first, see https://github.com/DLR-RM/stable-baselines3/issues/622 and related PR like https://github.com/DLR-RM/stable-baselines3/pull/1622

We need help there, especially for benchmarking and having efficient PER implementation that works with VecEnv.

**Comment by VIPTankz** [comment]
Hello there, I'm the original author on Beyond The Rainbow, but also have a Rainbow DQN implementation. This uses the same PER implementation as BTR, which can handle vectorized envs, n-step learning while only using around 12gb of RAM.

 In my implementation, there are a few minor differences from the original paper I'd like your thoughts on:
- **Parallel Envs and Batch Size:** Rainbow DQN used a single environment, therefore the batch size (32) and learning rate (6.25e-5) was based around this. In my implementation, I currently am using the BTR style vectorized environments, and the hyperparameters to match. I currrently have 64 environments, a batch size of 512 (this matches the ratio of steps to samples as in Rainbow DQN, where each transition is sampled 8 times), and a learning rate of 1e-4.
- **Terminal on life loss:**  The original Rainbow paper used Terminal on life loss, however this has since been advised against ("Revisiting the arcade learning environment: Evaluation protocols and open problems for general agents."). Currently I am using the more standard settings, where no terminal on life loss is used, nor is the agent given life information. This may however cause some performance discrepancy between my implementation and the original for games that use lives (this can be significant in some cases).
- **Evaluations:** My implementation also currently follows the evaluation protocol set out in the given paper, whereby every 250K steps (1M frames), an evaluation of 100 episodes is performed. This will however, slow down training considerably (9 hours to around 18 hours on my PC), even with my current implementation which does evaluation and training in parallel.

Please let me know what you think!

**Comment by araffin** [comment]
> What is that an implementation of exactly? 

this is DQN, if you compare it to [SB3 (PyTorch) run](https://wandb.ai/openrlbenchmark/sb3/runs/yiv1bqwd), it gives the same performance but is 10x faster.

I used it to be able to debug/test the PER implementation and iterate faster but I didn't manage to get both the speed and the correct performance so far.

> I can do this, but this will slow the training down significantly (probably at very least double walltime).

at first, to check how it impact performance, you could do 3 runs with standard hyperparams and 3 with the increased batch size/lr.
My main motivation is to be able to compare to the original PER/Rainbow papers and check that the results are matching.

**Comment by araffin** [comment]
>  Tyler has implemented Rainbow faithfully and we can continue refining the implementation and addressing issues that relate to correctness, stability and integration into the SB3 codebase.
>  is now close to a state where it can be pushed

Good to hear =)
Let's open a draft PR in the SB3 repository as soon as possible so I can provide early feedback and assistance.

> Optimising Rainbow's runtime through major re‑engineering, for example by reimplementing components in JAX or introducing substantial architectural changes, sits outside what we can take on at this stage.

I totally agree. My initial push for optimizing for speed was because I didn't have the resources to benchmark a slow algorithm.
The plan now is to first include a working (slow) version of RAINBOW in SB3 and then see if we can improve its speed (the main bottleneck in my previous trials was the replay buffer).

**Comment by roboserg** [comment]
6 month later, any progress on this? :)
